src/irr/video/COpenGLSpecializedShader.h

```diff
@@ -23,21 +25,48 @@ class COpenGLSpecializedShader : public IGPUSpecializedShader
         core::smart_refctd_dynamic_array<uint8_t> binary;
     };
 
-    COpenGLSpecializedShader(const asset::ICPUBuffer* _spirv, const asset::ISpecializationInfo* _specInfo, const asset::CIntrospectionData* _introspection);
+    COpenGLSpecializedShader(size_t _ctxCount, const asset::ICPUBuffer* _spirv, const asset::ISpecializationInfo* _specInfo, const asset::CIntrospectionData* _introspection);
 
-    //! @returns GL name or zero if already compiled once or there were compilation errors.
-    GLuint compile(uint32_t _GLSLversion);
+    inline GLuint getGLnameForCtx(const void* _ctx)
+    {
+        std::unique_lock<std::mutex> lock(m_getGLnameMutex);
+
+        auto found = std::find_if(m_GLnames.begin(), m_GLnames.end(), [&_ctx] (const CtxGLnamePair& n) { return n.first==_ctx; });
+        if (found != m_GLnames.end())
+            return found->second;
+
+        if (m_binary.binary) {
+            const GLuint GLname = COpenGLExtensionHandler::extGlCreateProgram();
+            COpenGLExtensionHandler::extGlProgramBinary(GLname, m_binary.format, m_binary.binary->data(), m_binary.binary->size());
+            m_GLnames.emplace_back(_ctx, GLname);
+            return GLname;
+        }
+
+        const GLuint GLname = compile(COpenGLExtensionHandler::ShaderLanguageVersion);
+        m_GLnames.emplace_back(_ctx, GLname);
+        return GLname;
+    }
 
     void setUniformsImitatingPushConstants(const uint8_t* _pcData, GLuint _GLname);
 
-    inline const SProgramBinary& getProgramBinary() const { return m_binary; }
+    inline GLenum getStage() const { return m_stage; }
 
-    GLenum getStage() const { return m_stage; }
+protected:
+    ~COpenGLSpecializedShader()
+    {
+        //shader programs can be shared so all names can be freed by any thread
+        for (auto& n : m_GLnames)
+            COpenGLExtensionHandler::extGlDeleteProgram(n.second);
+    }
 
 private:
+    //! @returns GL name or zero if already compiled once or there were compilation errors.
+    GLuint compile(uint32_t _GLSLversion);
     void buildUniformsList(GLuint _GLname);
 
 private:
+    using CtxGLnamePair = std::pair<const void*, GLuint>;
+    core::vector<CtxGLnamePair> m_GLnames;
```

Comment:
a plain and length-immutable `GLuint* mGLnames` would be more appropriate

Reply:
resolved